### ISSUE TITLE
Expose the .ForEach() and .Where() methods as regular PowerShell operators

### ISSUE THREAD

**Issue by mklement0** [PowerShell/PowerShell#6576]
Note:   
* This proposal arose out of the discussion in #6512  
* An RFC would have to be written, but I first want to see if there's a fundamental willingness to implement this.

---

The ~~hidden~~ and effectively undocumented `.Where()` and `.ForEach` _methods_ available on every object exposed by PowerShell were introduced to support DSC in v4 and provide functionality similar to the `Where-Object` and `ForEach-Object` cmdlets, but (a) in a more performant way (albeit in an all-in-memory way)  and (b) with more features. The most comprehensive documentation that I'm aware of is [this blog post by @KirkMunro](http://www.powershellmagazine.com/2014/10/22/foreach-and-where-magic-methods/).

While they could simple be documented (and made more discoverable), **I suggest surfacing them as bona fide PowerShell _operators_ `-foreach` and `-where`**:

* No _native_ PowerShell features  that I am aware of are exposed as _methods_: instead, functionality is surfaced as _commands_ (cmdlets/functions) and _operators_.

  * Methods belong to a _different realm_, to which [`Get-Help about_Methods`](https://github.com/PowerShell/PowerShell-Docs/blob/staging/reference/6/Microsoft.PowerShell.Core/About/about_Methods.md) is the portal (albeit one that, unfortunately, currently doesn't discuss syntax pitfalls and doesn't even reference [`Get-Help about_Parsing`](https://github.com/PowerShell/PowerShell-Docs/blob/staging/reference/6/Microsoft.PowerShell.Core/About/about_Parsing.md)); it's a powerful realm and definitely worth knowing about, but it is a different realm, distinct from how PowerShell itself surfaces its functionality.

* The `.Where()` and `.ForEach()` methods were introduced as such for the benefit of _DSC_ - ~~of necessity, I presume, given that the DSC DSL doesn't allow use of _pipelines_~~ (see comments below).

   * Despite their general availability inside PowerShell itself, they're effectively undocumented, and many people may not even be aware of their existence (since ~~tab completion doesn't help~~ (see comments below), and neither `Get-Member` nor `.psobject.Methods` find them, you simply have to _know_ of their existence).

   * Even if fully documented (and, as discussed, this is currently even lacking in the _DSC_ context, and overall Kirk's blog post is still the most complete documentation), these methods make awkward tools in general PowerShell use, due to their method syntax and output data types.

Hence my suggestion to **surface them as _operators_ with `[object[]]` output**, which would make them discoverable interactively (`-<tab>`) , and they would be documented alongside the existing operators.

They'd make **concise complements to their cmdlet counterparts**; `-foreach` would make the `foreach` _loop_ unnecessary in many situations and with its use of `$_` reduce the existing confusion between the loop and the cmdlet (see https://github.com/PowerShell/PowerShell-Docs/issues/1514 and a rejected attempt to harmonize the two, #3830):

```powershell
$var = 1, 2, 3 -foreach { $_ + 1 }  # wishful thinking

# vs.

$var = foreach ($i in 1, 2, 3) { $i + 1 }
# or (slower)
$var = 1, 2, 3 | ForEach-Object { $_ + 1 }

#  ---

$var = 1, 2, 3 -where { $_ -gt 1 } # wishful thinking

# vs.
# (slower)
$var = 1, 2, 3 | Where-Object { $_ -gt 1 }
# or
$var = foreach ($i in 1, 2, 3)  { if ($i -gt 1) { $i } }

```

Arguably, even DSC could benefit from the cleaner operator syntax; contrast the following:

```powershell
# Note that this example, taken from https://docs.microsoft.com/en-us/powershell/dsc/separatingenvdata
# doesn't even use `()` around the script-block argument, so the use of Where() 
# isn't even readily recognizable as a *method* call.
Node ($AllNodes.Where{$_.Role -eq "WebServer"}).NodeName { ...

# vs.

Node ($AllNodes -where {$_.Role -eq "WebServer"}).NodeName { ...  # wishful thinking
```

Environment data
----------------

Written as of:

```powershell
PowerShell Core v6.0.2
```

**Comment by BrucePay** [comment]
@mklement0 Please open an RFC. I think it will be much easier to manage commenting through that mechanism.

**Comment by BrucePay** [comment]
@mklement0  Some (ok many) comments. But please proceed with the RFC.

> The most comprehensive documentation that I'm aware of is this blog post by @KirkMunro.

Please open a doc bug for .foreach{} and .where{}.

> No native PowerShell features that I am aware of are exposed as methods

Methods **_are_** a native PowerShell feature, supported directly in the language. At one point we had something like
```
Invoke-Member $obj memberName arg1, arg2, arg3
```
but decided to have native support in the language because we knew we'd never get all the coverage we wanted.

>Methods belong to a different realm, to which Get-Help about_Methods is the portal (albeit one that, unfortunately, currently doesn't discuss syntax pitfalls and doesn't even reference Get-Help about_Parsing); 

What pitfalls and why should it reference about_Parsing?

> The .Where() and .ForEach() methods were introduced as such for the benefit of DSC 
 
While the proximate cause for adding .where{} and .foreach{} was DSC configuration data, they are general purpose functionality. 

> of necessity, I presume, given that the DSC DSL doesn't allow use of pipelines.

This is incorrect.

> Despite their general availability inside PowerShell itself, they're effectively undocumented, 

Again, please open a doc bug.

> and many people may not even be aware of their existence (since tab completion doesn't help, and neither Get-Member nor .psobject.Methods find them, you simply have to know of their existence).

Tab-completion does complete these methods.
```
PS[1] (32) > (1,2,3).F<tab>
PS[1] (32) > (1,2,3).ForEach(
```
>They'd make concise complements to their cmdlet counterparts; -foreach would make the foreach loop unnecessary in many situations and with its use of $_ reduce the existing confusion between the loop and the cmdlet - see #3830:

I'd actually be concerned about _increasing_ the confusion considering that the user can do either
```
Get-Process -where { $_.CPU -gt 100}
```
or
```
Get-Process |where { $_.CPU -gt 100 }
```
which appear to do the same thing but doesn't (quite) and whose notation differs only by a single character. How do you explain that to a novice user?

Finally, in the last examples, there are unnecessary parens in the `.where{}` case. The corrected example is
```
Node $AllNodes.Where{$_.Role -eq "WebServer"}.NodeName 
{ ...
```
And again your proposed alternative example
```
Node ($AllNodes -where {$_.Role -eq "WebServer"}).NodeName
{ ... 
```
can just be written as
```
Node ($AllNodes |where {$_.Role -eq "WebServer"}).NodeName
{ ... 
```
Given the only syntactic difference is '|' before `where` instead of '-', I don't see an advantage in _this_ scenario. So - going back to basics, - the scenario that `.where{}` was intended to address was the deep extraction of data using compact expressions with minimal syntactic noise. For example
```
$procs.where{ $_.cpu -gt 100 }.modules.where{ $_.modulename -match "ntdll"}.count # zero parens
```
or
```
write-output $n.foreach{ $_ * 2 } 
```
Writing the first example using `-foreach` would look something like
```
(((gps) -where { $_.cpu -gt 100 }).modules -where{ $_.modulename -match "ntdll"}).count
```
This has a lot of nested parens and that is exactly what we were trying to get away from.

So - from an RFC perspective I'd be looking for at least:

- target audience (novice, advanced, ...)
- compelling scenarios (how this makes PowerShell better e.g. opening up new scenarios, making existing ones more efficient, ...)
- significant examples where these scenarios are addressed (we had this conversation in the committee meeting this week - too few of the RFC's have significant (or any!) examples.)
- risks (?)
-

**Comment by mklement0** [comment]
> Tab-completion does complete these methods.

Thanks for clarifying that. I had tried with a _scalar_, where it doesn't work: `$v = 1; $v.f<tab>`

On the command line, tab completion decides based on the whether the variable value is array-valued whether to offer the method.

By contrast, the VSCode extension isn't smart enough to detect _potentially_ array-valued variables, such as `$v = Get-ChildItem; $v.f<tab>`

Arguably, tab completion of these methods should be offered on both scalars and collections, because they truly are available on both - that would also simplify the logic.

Either way, these problems are certainly surmountable.

**Comment by mklement0** [comment]
> I'd actually be concerned about increasing the confusion considering that the user can do either
> 
> Get-Process -where { $_.CPU -gt 100}
> or
> 
> Get-Process |where { $_.CPU -gt 100 }
>
> which appear to do the same thing but doesn't (quite) and whose notation differs only by a single 
> character. 

More than 1 character difference: your first example wouldn't work, because you're missing the `(...)` around `Get-Process`:

```powershell
 # Get-Process needs (...) to allow its output to participate in the expression
(Get-Process) -where { $_.CPU -gt 100}  
```

> How do you explain that to a novice user?

The same way you explain the existing difference between `... | foreach { ... }` and `foreach (...) { ... }`:

Ideally, create a new conceptual help topic that discusses the tradeoffs between those constructs:

* pipeline syntax: one-by-one as-output-is-produced processing: memory-throttling, but slow

* operator syntax: up-front output-collecting, then iterating: memory-intensive, but fast

Even without introducing the proposed operator these tradeoffs are worth documenting with respect to `... | foreach { ... }` vs. `foreach (...) { ... }`.

The new `-foreach` operator would have the added advantage of eliminating the confusion around `$_` use in `foreach` loops.

From a syntax perspective, the `-foreach` vs. `... | foreach` dichotomy strikes me as cleaner and easier to explain. (The `foreach` _loop_ will still have its place for more elaborate loops.)

**Comment by mklement0** [comment]
> deep extraction of data using compact expressions with minimal syntactic noise. For example
> `$procs.where{ $_.cpu -gt 100 }.modules.where{ $_.modulename -match "ntdll"}.count # zero parens`

Again, I wouldn't drop the method-call parentheses:

```powershell
$procs.where({ $_.cpu -gt 100 }).modules.where({ $_.modulename -match "ntdll"}).count
```

vs. 

```powershell
(($procs -where { $_.cpu -gt 100 }).modules -where { $_.modulename -match "ntdll"}).count
```

Ultimately the same number of parentheses (though the nesting adds complexity), and also whitespace to clarify the structure, and the parentheses only come into play when property / method access is needed as well.

A fluent API (method chaining) certainly has its charms, but it doesn't strike me as PowerShell-like.
That said, people who wish to use it still can.

**Comment by mklement0** [comment]
> Methods _are_ a native PowerShell feature, supported directly in the language. At one point we had 
> something like
> 
> `Invoke-Member $obj memberName arg1, arg2, arg3`
> but decided to have native support in the language because we knew we'd never get all the coverage we wanted.

Yes, and the direct access to methods is great.

But there is a difference between _enabling_ direct access to features _from a different realm_ (.NET) and PowerShell's _own_ of surfacing functionality (commands and operators).

An example would be PowerShell's `-split` operator as a (superior) analog to `String.Split()` - yes, you  _can_ call the latter, but why would you? There's the familiar syntax of a PowerShell operator, the implicit conversion of the LHS to string, the support for regexes, ...

Similarly, `.Where()` and `.ForEach()` - as unique PowerShell features that act _like_ operators - _should_ be operators.

**Comment by vexx32** [comment]
I think this is likely to be a good idea, even if only to reduce the syntactic irregularities that `.ForEach({})` and `.Where({})` introduce. Additionally, if `-foreach` were to be implemented, I would further recommend that `| foreach {}` be *removed* as an alias for `ForEach-Object`. 

The presence of `| foreach {}` alongside `foreach($obj in $col) {}` is the source of a lot of confusion for beginners, and even some more advanced users who aren't aware of the `| foreach {}` alias.

The alternative `-foreach {}` is a significantly simpler syntax, but I must wonder if the intent is to *replace* `.ForEach({})` then we must have some way of supporting the multiple parameters of `.ForEach({})` with the operator, and from what I've seen I don't think any operators typically extend beyond unary or binary -- having support for something like `$collection -foreach {} {} {}` would be very unusual indeed, and potentially be quite syntactically confusing.

I'd think it might be best to have `-foreach` only support the one script block argument.

**Comment by vexx32** [comment]
Those are still binary operators, though; their second operand is simply an array with optional elements.

What I was thinking about is how `ForEach-Object` takes input like this: `$array | ForEach-Object {"Begin"} {"Process"} {"End"}` -- which it does by having them assigned to separate parameters. I suppose a similar thing could be done with a `-foreach {'begin'},{'process'},{'end'}` but personally I prefer to avoid overcomplicating the syntax. Some of PS's best stuff is in its clarity, and this would... not be super clear, I feel.

**Comment by mklement0** [comment]
@vexx32: 

> Those are still binary operators, though; their second operand is simply an array with optional elements.

Absolutely.

> What I was thinking about is how ForEach-Object takes input like this: $array | ForEach-Object {"Begin"} {"Process"} {"End"}

This mostly makes sense in a _pipeline_ context - I agree that it's not needed in an _operator_ context.

**Comment by mklement0** [comment]
@BrucePay: I've created the following docs issues:

* https://github.com/PowerShell/PowerShell-Docs/issues/2307
* https://github.com/PowerShell/PowerShell-Docs/issues/2308

On a meta note: I've griped about the state of PowerShell's documentation before, but it truly is disheartening whenever I take a closer look. Let's hope that will change for the better with help from the community.

I will create an RFC soonish.
